crud/src/main/java/com/redhat/lightblue/crud/TranslatorToJson.java

```diff
@@ -0,0 +1,185 @@
+        String fieldName = field.getName();
+        Path path = fieldCursor.getCurrentPath();
+
+        Error.push(field.getFullPath().getLast());
+
+        try{
+            JsonNode newJsonNode = null;
+            Object value = getValueFor(source, path);
+
+            if (field instanceof ObjectField) {
+                newJsonNode = translate(source, (ObjectField)field, fieldCursor);
+            }
+            else if(value != null){
+                if (field instanceof SimpleField) {
+                    newJsonNode = translate((SimpleField)field, value);
+                }
+                else if (field instanceof ArrayField){
+                    newJsonNode = translate((ArrayField)field, value, fieldCursor);
+                }
+                else if (field instanceof ReferenceField) {
+                    newJsonNode = translate((ReferenceField)field, value);
+                }
+                else{
+                    throw new UnsupportedOperationException("Unknown Field type: " + field.getClass().getName());
+                }
+            }
+
+            targetNode.set(fieldName, newJsonNode);
+        }
+        finally{
+            Error.pop();
+        }
+    }
+
+    protected JsonNode translate(S source, ObjectField field, FieldCursor fieldCursor){
+        if(!fieldCursor.firstChild()){
+            //TODO: Should an exception be thrown here?
+            return null;
+        }
+
+        ObjectNode node = factory.objectNode();
+
+        iterateOverNodes(source, node, fieldCursor);
+
+        fieldCursor.parent();
+
+        return node;
+    }
+
+    protected JsonNode translate(ArrayField field, Object o, FieldCursor fieldCursor){
+        if(!fieldCursor.firstChild()){
+            //TODO: Should an exception be thrown here?
+            return null;
+        }
+
+        FieldTreeNode node = fieldCursor.getCurrentNode();
+
+        ArrayElement arrayElement = field.getElement();
+        ArrayNode valueNode = factory.arrayNode();
+
+        List<? extends Object> values;
+        if (arrayElement instanceof SimpleArrayElement) {
+            values = getSimpleArrayValues(o);
+        }
+        else if(arrayElement instanceof ObjectArrayElement){
+            values = getObjectArrayValues(o);
+        }
+        else{
+            throw new UnsupportedOperationException("ArrayElement type is not supported: " + node.getClass().getName());
+        }
+
+        for(Object value : values){
+            valueNode.add(toJson(node.getType(), value));
+        }
+
+        fieldCursor.parent();
+        return valueNode;
+    }
+
+    protected abstract JsonNode translate(ReferenceField field, Object o);
+    protected abstract JsonNode translate(SimpleField field, Object o);
+
+    protected abstract Object getValueFor(S source, Path path);
+    protected abstract List<? extends Object> getSimpleArrayValues(Object o);
+    protected abstract List<? extends Object> getObjectArrayValues(Object o);
```

Comment:
> getObjectArrayValues will need the cursor to get the metadata fields, and parse the object entries. Without that, there is no way it can know what fields it needs to parse.

I don't actually have a use case for this scenario, but I think I can take a stab at it.

> getSimpleArrayValues needs at least the SimpleArrayElement object, so it can know the type of the value it is processing.

What about just passing in the Type itself?